### PR TITLE
Unbounded foreach

### PR DESCRIPTION
**Pull request by romain-intel**

### CODE REVIEW

File: metaflow/runtime.py [hunk 1]
```diff
@@ -337,14 +439,17 @@ def _queue_tasks(self, finished_tasks):
             self._is_cloned[task.path] = task.is_cloned
 
             # CHECK: ensure that runtime transitions match with
-            # statically inferred transitions
+			# statically inferred transitions. Make an exception for control
```
Comment (romain-intel) [inline metaflow/runtime.py:442]: Nit: indentation issue.

File: metaflow/runtime.py [hunk 2]
```diff
@@ -508,13 +635,28 @@ def __init__(self,
...
                 # determine the number of retries of this task
                 user_code_retries, error_retries = deco.step_task_retry_count()
-                self.user_code_retries = max(self.user_code_retries,
-                                             user_code_retries)
-                self.error_retries = max(self.error_retries, error_retries)
+                if user_code_retries is None and error_retries is None:
+                    # This signals the runtime that the task doesn't want any
+                    # retries indifferent to other decorator opinions.
+                    # NOTE: We don't statically disallow specifying `@retry` in
+                    # combination with other decorators (e.g. `@archer`) only to
```
Comment (romain-intel) [inline metaflow/runtime.py:647]: Internal comment needs rewording.

File: metaflow/runtime.py [hunk 3]
```diff
@@ -799,7 +933,8 @@ def _launch(self):
             for deco in self.task.decos:
                 deco.runtime_step_cli(args,
                                       self.task.retries,
-                                      self.task.user_code_retries)
+                                      self.task.user_code_retries,
+									  self.task.ubf_context)
```
Comment (romain-intel) [inline metaflow/runtime.py:937]: Nit: Indentation

File: metaflow/util.py [hunk 4]
```diff
@@ -290,8 +290,7 @@ def dict_to_cli_options(params):
             if k == 'decospecs':
                 k = 'with'
             k = k.replace('_', '-')
-            if not isinstance(v, tuple):
-                v = [v]
+            v = v if isinstance(v, list) or isinstance(v, tuple) else [v]
             for value in v:
                 yield '--%s' % k
                 if not isinstance(value, bool):
```
Comment (romain-intel) [inline metaflow/util.py:296]: Internal PR had a to_unicode here. Is this no longer relevant?

**Review by romain-intel — COMMENTED** [review]
A few comments mostly related around cleanup. LGTM otherwise.

**Comment by romain-intel** [comment]
@seeravikiran: could you make the small changes suggested and then we can merge if that's ok.

**Comment by savingoyal** [comment]
@romain-intel Is this PR linked to datastore changes or can I go ahead and test this PR on top of master?

File: metaflow/plugins/__init__.py [hunk 5]
```diff
@@ -61,7 +63,8 @@ def _merge_lists(base, overrides, attr):
                                 RetryDecorator,
                                 BatchDecorator,
                                 StepFunctionsInternalDecorator,
-                                CondaStepDecorator], ext_plugins.STEP_DECORATORS, 'name')
+                                CondaStepDecorator,
+                                TestUnboundedForeachDecorator], ext_plugins.STEP_DECORATORS, 'name')
```
Comment (savingoyal) [inline metaflow/plugins/__init__.py:67]: I don't think we should expose this decorator right now.

**Review by savingoyal — CHANGES_REQUESTED** [review]

**Comment by romain-intel** [comment]
I think you can port this PR on top of master. I don't think it has a strong dependency (but things will change a bit).

**Comment by romain-intel** [comment]
Replaced by #554

**Pull request closed without merging**